### PR TITLE
Adding image driver and restructuring

### PR DESCRIPTION
**Pull request by jsignell**
This PR adds an image reader for `xarray` and does a restructure to have a common structure for all plugins in `intake-xarray`. In particular this adds the ability to return a dataset rather than a dataarray using the `merge_dim` option. 

I think this will help justify the existence of an intake-xarray plugin at all since it can now be used directly and the other plugins are just helper for using specific readers. 

It is likely that this needs lots of work. I rewrote the example notebook to try to explain some of the behavior. 

screen shot 2018-12-13 at 1 52 15 pm

### CODE REVIEW

**Comment by martindurant** [comment]
Are you around the coming week to take me through the proposal here?

**Comment by philippjfr** [comment]
Just tried playing with this, but I can't figure out why ``intake.open_image`` doesn't exist for me.

**Comment by martindurant** [comment]
@philippjfr , I believe @jsignell is not yet back from holidays. Maybe you get an explicit error when trying to import ImageSource?

**Comment by philippjfr** [comment]
Thanks for letting me know. No error when I try to import ``ImageSource`` directly. At some point I should look clearly read some of the internals of the plugin system to understand how the function gets registered.

**Comment by jsignell** [comment]
@martindurant I forgot to mention in our chat that this PR also adds the ability to return datasets using the kwarg `merge_dim` rather than only dataarrays. I am not sure if that is an overstep for intake or not, but it does seem like a bit of munging that it would be very handy to be able to encode in a catalog.

**Comment by martindurant** [comment]
I would say that adding extra capabilities that may be useful for some is totally in scope, so long as it doesn't ass complexity for those that don't need it.

File: intake_xarray/base.py [hunk 1]
```diff
@@ -1,38 +1,231 @@
...
+            k: xr.concat(
+                [xr.DataArray(
+                    np.full(das[i].sizes.get(self.concat_dim, 1), v),
+                    dims=self.concat_dim
+                ) for i, v in enumerate(values) if index is None or i in index],
+                dim=self.concat_dim)
+            for k, values in field_values.items() if k != self.merge_dim
+        }
+
+    def _open_files(self, files):
```
Comment (jsignell) [inline intake_xarray/base.py]: A lot of logic is in this method. Not sure how to split it up better though. Essentially there are 4 paths through, no pattern and merge_dim, no pattern and concat_dim, pattern and concat_dim, or pattern and merge_dim. The first two are pretty straightforward and then they increase in complexity.

File: intake_xarray/base.py [hunk 2]
```diff
@@ -1,38 +1,231 @@
...
+        super(XarraySource, self).__init__(metadata=metadata)
+
+    def reader(self, filename, chunks, **kwargs):
+        if getattr(self, '_reader', None) is not None:
+            return self._reader(filename, chunks, **kwargs)
+        raise NotImplementedError('Plugin must implement a reader function '
+                                  'which takes at least filename, and chunks '
+                                  'and returns an xarray object')
+
+    def multireader(self, filename, chunks, **kwargs):
```
Comment (martindurant) [inline intake_xarray/base.py]: Also

File: intake_xarray/base.py (hunk 1 above)
Comment (martindurant) [reply intake_xarray/base.py]: Doesn't look too hairy. Perhaps could do with a comment on each branch, saying what it does, and a docstring with your comment above

File: intake_xarray/base.py [hunk 3]
```diff
@@ -1,38 +1,231 @@
...
+                    dims=self.concat_dim
+                ) for i, v in enumerate(values) if index is None or i in index],
+                dim=self.concat_dim)
+            for k, values in field_values.items() if k != self.merge_dim
+        }
+
+    def _open_files(self, files):
+        das = [self.reader(f, chunks=self.chunks, **self.kwargs)
+               for f in files]
+        if not self.pattern:
```
Comment (martindurant) [inline intake_xarray/base.py]: Perhaps some idea of which of these attributes are mutually exclusive

File: intake_xarray/base.py [hunk 4]
```diff
@@ -1,38 +1,231 @@
...
+                coords = {k: values[i] for k, values in field_values.items() if k != self.merge_dim}
+                da = da.assign_coords(**coords).chunk(self.chunks)
+            if isinstance(da, xr.DataArray):
+                da.name = data_var
+            merge_das.append(da)
+
+        return xr.merge(merge_das)
+
+    def _open_dataset(self):
+        from glob import glob
```
Comment (martindurant) [inline intake_xarray/base.py]: We are assuming local files here? Do any readers work more generally? 
Note that open_files (if you can use any file-like objects) or get_fs_token_paths (if you need a set of string paths) already handle list and glob expansion to give you one or more files, you probably don't need to duplicate here.

Drivers that do indeed work only with local files should probably have some text in the docs and/or docstrings about how to make them work with caching from remote.

File: intake_xarray/base.py [hunk 5]
```diff
@@ -1,38 +1,231 @@
...
+            self._ds = self._open_files(self.urlpath)
+        else:
+            self._ds = self.reader(self.urlpath, chunks=self.chunks,
+                                   **self.kwargs)
 
     def _get_schema(self):
         """Make schema object, which embeds xarray object and some details"""
         from .xarray_container import serialize_zarr_ds
 
+        is_multi = '*' in self.urlpath or isinstance(self.urlpath, list)
```
Comment (martindurant) [inline intake_xarray/base.py]: Because you might pass the glob through without processing?

File: intake_xarray/base.py [hunk 6]
```diff
@@ -1,38 +1,231 @@
...
         self.urlpath = self._get_cache(self.urlpath)[0]
+        if isinstance(self.urlpath, list) and not is_multi:
+            self.urlpath = self.urlpath[0]
```
Comment (martindurant) [inline intake_xarray/base.py]: `assert len(self.urlpath) == 1` ?

File: intake_xarray/base.py [hunk 7]
```diff
@@ -1,38 +1,231 @@
...
 
         if self._ds is None:
             self._open_dataset()
 
+            if isinstance(self._ds, xr.DataArray):
+                da = self._ds
+                ds = xr.Dataset({'value': da})
+                extra_metadata = {'array': 'value'}
+            else:
+                ds = self._ds
```
Comment (martindurant) [inline intake_xarray/base.py]: comment: if not an array, it must be a DataSet

File: intake_xarray/base.py [hunk 8]
```diff
@@ -1,38 +1,231 @@
...
+                metadata['internal'] = serialize_zarr_ds(ds)
+
             self._schema = Schema(
                 datashape=None,
-                dtype=None,
-                shape=None,
-                npartitions=None,
+                dtype=str(da.dtype),
+                shape=da.shape,
+                npartitions=getattr(da.data, 'npartitions', None),
```
Comment (martindurant) [inline intake_xarray/base.py]: This is populated only for the case of an array? I assume that, in general, the number of chunks can be different for each variable of a data-set.

File: intake_xarray/image.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
+import xarray as xr
```
Comment (martindurant) [inline intake_xarray/image.py]: To be deferred

File: intake_xarray/image.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+
+def dask_imread(filename, imread=None, preprocess=None):
+    """ Read a stack of images into a dask array
+    Parameters
+    ----------
+    filename: string, iter
+        A globstring like 'myfile.*.png', a list of files, or anything
+        that can go into imread.
+    imread: function (optional)
```
Comment (martindurant) [inline intake_xarray/image.py]: Do we support passing functions through? If yes, probably need some careful documentation on how to do that, and how to express it in a catalogue (forgive me, if this is already in the doc/example). If not, can simplify this function.

File: intake_xarray/image.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    >>> from dask.array.image import imread
+    >>> im = imread('2015-*-*.png')  # doctest: +SKIP
+    >>> im.shape  # doctest: +SKIP
+    (365, 1000, 1000, 3)
+    Returns
+    -------
+    Dask array of all images stacked along the first dimension.  All images
+    will be treated as individual chunks
+    """
+    import os
```
Comment (martindurant) [inline intake_xarray/image.py]: Comment: copied from ...

File: intake_xarray/image.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    from dask.array import Array
+    from dask.base import tokenize
+
+    imread = imread or sk_imread
+    name = None
+
+    if isinstance(filename, list):
+        filenames = filename
+    elif '*' in filename:
+        filenames = sorted(glob(filename))
```
Comment (martindurant) [inline intake_xarray/image.py]: ks_imread definitely can read from arbitrary file-like objects, so this case should certainly use open_files, so we can directly access s3, gcs... . The only wrinkle is, that the files are not really opened until coming to a `with` block, so you'll need a little helper function like

```
def read_open_file(open_file, imread):
    with open_file as f:
        return imread(f)
```
and we wish to specify, if the user is to provide their own function, that it must either take any file-like

File: intake_xarray/image.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    elif os.path.isfile(filename):
+        filenames = [filename]
+    else:
+        filenames = [filename]
```
Comment (martindurant) [inline intake_xarray/image.py]: What is the expected use-case here? We want to allow passing a directory?

File: intake_xarray/image.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    """Open a xarray dataset from image files.
+
+    This creates an xarray.DataArray or an xarray.Dataset.
+
+    See http://scikit-image.org/docs/dev/api/skimage.io.html#skimage.io.imread
+    for the file formats supported, and
+     http://docs.dask.org/en/latest/array-api.html#dask.array.image.imread
+    for possible extra arguments.
+
+    NOTE: Although ``skimage.io.imread`` is used by default, any reader function which
```
Comment (martindurant) [inline intake_xarray/image.py]: Should somewhere give an example of how that works.

Comment (martindurant) [reply intake_xarray/image.py]: (or just remove the capability??)

File: intake_xarray/netcdf.py [hunk 15]
```diff
@@ -1,33 +1,39 @@
...
+class NetCDFSource(XarraySource):
+    """Open a netcdf file.
+
+    This creates an xarray.Dataset.
+
+    See http://xarray.pydata.org/en/stable/generated/xarray.open_dataset.html
+    for the file formats supported and possible extra arguments.
+
+    NOTE: When reading from OpenDAP URLs do not set the ``chunks`` option to
+    use provided default chunking.
```
Comment (martindurant) [inline intake_xarray/netcdf.py]: There is an explanation under the opendap driver that it handles the auth part - should say here explicitly and that the other driver may be necessary

File: intake_xarray/netcdf.py [hunk 16]
```diff
@@ -1,33 +1,39 @@
...
+    urlpath : str
+        Path to source file. May include glob "*" characters or format
+        pattern strings. Accepts un-authenticated OpenDAP urls, and any
+        remote source can be used as long as caching is enabled.
+        Must be a format supported by ``xr.open_dataset``.
+
+        Some examples:
+            - ``s3://data/*.nc``
+            - ``http://thredds.ucar.edu/thredds/dodsC/grib/FNMOC/WW3/Global_1p0deg/Best``
+            - ``https://github.com/pydata/xarray-data/blob/master/air_temperature.nc?raw=true``
```
Comment (martindurant) [inline intake_xarray/netcdf.py]: So does this handle remote URLs directly or not? I assume if it is opendap, then yes, in which case the thing about needing caching is wrong (and in fact won't work).
The thredds URL gives Unrecognized Request for me.

File: tests/test_intake_xarray.py [hunk 17]
```diff
@@ -16,7 +16,7 @@ def test_discover(source, cdf_source, zarr_source, dataset):
     r = source.discover()
 
     assert r['datashape'] is None
-    assert r['dtype'] is None
+    assert r['dtype'] == 'float32'
```
Comment (martindurant) [inline tests/test_intake_xarray.py]: The whole dataset has a single dtype?

File: intake_xarray/base.py [hunk 18]
```diff
@@ -1,38 +1,231 @@
...
+        self.concat_dim = concat_dim
+        self.merge_dim = merge_dim
+        self.kwargs = xarray_kwargs or kwargs
+        self._ds = None
+        self.storage_options = storage_options
+        self._reader = reader or xr.open_dataset
+        self._multireader = multireader or xr.open_mfdataset
+        super(XarraySource, self).__init__(metadata=metadata)
+
+    def reader(self, filename, chunks, **kwargs):
```
[the remark this answers is not in the document]
Comment (jsignell) [reply intake_xarray/base.py]: I was trying to set a default while still enforcing that plugins define the reader and multireader. But maybe this isn't the right way...

File: intake_xarray/base.py (hunk 6 above)
Comment (jsignell) [reply intake_xarray/base.py]: :+1:

File: intake_xarray/base.py (hunk 4 above)
Comment (martindurant) [reply intake_xarray/base.py]: `dask.bytes.open_files` (also `fsspec.open_files`), i.e., the generic way to open anything, local or remote, glob or list

File: intake_xarray/image.py (hunk 14 above)
Comment (martindurant) [reply intake_xarray/image.py]: I think if it's that is importable, it's easy to do: `!!python/name:mymodule.process`. Could be seen as a future enhancement?

File: intake_xarray/base.py [hunk 19]
```diff
@@ -1,38 +1,231 @@
 from . import __version__
 from intake.source.base import DataSource, Schema, PatternMixin
-from .xarray_container import ZarrSerialiser
+from intake.source.utils import reverse_formats
 
+try:
+    import xarray as xr
```
[the remark this answers is not in the document]
Comment (martindurant) [reply intake_xarray/base.py]: In the case that xarray is available, it'll make the import of intake slower. In the case it isn't, the module won't load, but the exception will get swallowed.
If deferred, the module would load OK, but when the user tries to access the data, *then* they'll get the message, saying that they need to install something if they want to use that source.

File: intake_xarray/image.py (hunk 14 above)
Comment (jsignell) [reply intake_xarray/image.py]: That seems reasonable.

**Comment by jsignell** [comment]
After conversation with Martin this chunk of work seems unreasonably big. So I am going to make a new PR that just adds an ImagePlugin.

**Comment by martindurant** [comment]
Closed in favour of #28 . The refactor may become necessary again at a later stage.

**Pull request closed without merging**